src/guides/anatomy-of-an-http-transaction.md

```diff
@@ -0,0 +1,410 @@
+
+    // END OF NEW STUFF
+  });
+}).listen(80);
+```
+
+## Echo Example
+
+In this example, all we're going to do is make a simple echo server, which just
+sends whatever data is received in the request right back in the response.
+
+```javascript
+var http = require('http');
+
+http.createServer(function(request, response) {
+  var body = '';
+  request.on('data', function(chunk) {
+    body += chunk;
+  }).on('end', function() {
+    response.end(body);
+  });
+}).listen(80);
+```
+
+Now let's tweak this a bit. Now we want to only send an echo under the following
+conditions:
+
+* The request method is GET.
+* The URL is `/echo`.
+
+In any other case, we want to simply respond with a 404.
+
+```javascript
+var http = require('http');
+
+http.createServer(function(request, response) {
+  if (request.method === 'GET' && request.url === '/echo') {
+    var body = '';
+    request.on('data', function(chunk) {
+      body += chunk;
+    }).on('end', function() {
+      response.end(body);
+    })
+  } else {
+    response.statusCode = 404;
+    response.end();
+  }
+}).listen(80);
+```
+
+> **Note:** By checking the URL in this way, we're doing a form of "routing".
+Other forms of routing can be as simple as `switch` statements or as complex as
+whole frameworks like [`express`][]. If you're looking for something that does
+routing and nothing else, try [`router`][].
+
+Great! Now let's take a stab at simplifying this. Remember, the `request` object
+is a [`ReadableStream`][] and the `response` object is a [`WriteableStream`][].
+That means we can use [`pipe`][] to direct data from one to the other. That's
+exactly what we want for an echo server!
+
+```javascript
+var http = require('http');
+
+http.createServer(function(request, response) {
+  if (request.method === 'GET' && request.url === '/echo') {
+    request.pipe(response);
+  } else {
+    response.statusCode = 404;
+    response.end();
+  }
+}).listen(80);
+```
+
+Yay streams!
+
+We're not quite done yet though. As mentioned multiple times in this guide,
+errors can and do happen, and we need to deal with them.
+
+To handle errors on the request stream, we'll log the error to `stderr` and send
+a 400 status code to indicate a `Bad Request`. In a real-world application,
+though, we'd want inspect the error to figure out what the correct status code
+and message would be. As usual with errors, you should consult the
+[`Error` documentation][].
+
+On the response, we'll just log the error to `stdout`.
+
+```javascript
+var http = require('http');
+
+http.createServer(function(request, response) {
+  request.on('error', function(err) {
+    console.error(err);
+    response.statusCode = 400;
+    response.end();
+  });
+  response.on('error', function(err) {
+    console.error(err);
+  });
+  if (request.method === 'GET' && request.url === '/echo') {
+    request.pipe(response);
+  } else {
+    response.statusCode = 404;
+    response.end();
+  }
+}).listen(80);
+```
+
```

Comment:
It might be good to have a quick "summing up what we learned and built" paragraph here, with links to further reading (for the moment, the API docs for event emitters, streams, and http.)